File: scripts/backend-node/idempotency.js
// Minimal in-process idempotency cache for the reference create endpoints.
//
// PRODUCTION NOTE: this is in-memory and per-process — fine for the demo, NOT
// fine for a real merchant deployment. Replace with Redis / DynamoDB / a DB
// table keyed on (idempotencyKey -> { status, body, createdAt }) before going
// live. The interface here is deliberately small so the swap is straightforward.

const TTL_MS = 24 * 60 * 60 * 1000;          // 24h
const MAX_ENTRIES = 10_000;
const cache = new Map();                       // key -> { status, body, createdAt }

function pruneIfNeeded() {
  if (cache.size <= MAX_ENTRIES) return;
  const cutoff = Date.now() - TTL_MS;
  for (const [k, v] of cache) {
    if (v.createdAt < cutoff) cache.delete(k);
    if (cache.size <= MAX_ENTRIES) break;
  }
}

export function getCached(key) {
  if (!key) return null;
  const entry = cache.get(key);
  if (!entry) return null;
  if (Date.now() - entry.createdAt > TTL_MS) {
    cache.delete(key);
    return null;
  }
  return entry;
}

export function setCached(key, status, body) {
  if (!key) return;
  pruneIfNeeded();
  cache.set(key, { status, body, createdAt: Date.now() });
}

/**
 * Read the idempotency key from headers OR body.
 * Header `Idempotency-Key` is the recommended convention; some clients pass
 * `idempotencyKey` in the JSON body — accept both for ergonomics.
 */
export function readKey(req) {
  const fromHeader = req.headers["idempotency-key"];
  if (typeof fromHeader === "string" && fromHeader.trim()) return fromHeader.trim();
  const fromBody = req.body?.idempotencyKey;
  if (typeof fromBody === "string" && fromBody.trim()) return fromBody.trim();
  return null;
}

File: scripts/backend-python/idempotency.py
"""Minimal in-process idempotency cache for the reference create endpoints.

PRODUCTION NOTE: this is in-memory and per-process — fine for the demo, NOT
fine for a real merchant deployment. Replace with Redis / DynamoDB / a DB row
keyed on (idempotency_key -> {status, body, created_at}) before going live.
The interface here is deliberately small so the swap is straightforward.
"""
import threading
import time
from typing import Optional

TTL_SECONDS = 24 * 60 * 60
MAX_ENTRIES = 10_000

_cache: dict = {}                      # key -> (status, body, created_at)
_lock = threading.Lock()


def _prune_locked():
    if len(_cache) <= MAX_ENTRIES:
        return
    cutoff = time.time() - TTL_SECONDS
    for k in list(_cache.keys()):
        if _cache[k][2] < cutoff:
            del _cache[k]
        if len(_cache) <= MAX_ENTRIES:
            break


def get_cached(key: Optional[str]):
    if not key:
        return None
    with _lock:
        entry = _cache.get(key)
        if not entry:
            return None
        if time.time() - entry[2] > TTL_SECONDS:
            del _cache[key]
            return None
        status, body, _ = entry
        return status, body


def set_cached(key: Optional[str], status: int, body) -> None:
    if not key:
        return
    with _lock:
        _prune_locked()
        _cache[key] = (status, body, time.time())


def read_key(headers, body) -> Optional[str]:
    """Header `Idempotency-Key` is the convention; some clients pass
    `idempotencyKey` in the JSON body — accept both for ergonomics."""
    h = headers.get("Idempotency-Key") or headers.get("idempotency-key")
    if isinstance(h, str) and h.strip():
        return h.strip()
    if isinstance(body, dict):
        b = body.get("idempotencyKey")
        if isinstance(b, str) and b.strip():
            return b.strip()
    return None
